### PR TITLE
Add interactive prompts for deployment name and work pool to `prefect deploy`

### PR DESCRIPTION
**Pull request by desertaxle**
Adds interactive prompts to the `prefect deploy` command to guide users through creating a deployment. The aim is to make it easier for users to create their first deployment by allowing `prefect deploy` to run only with an entrypoint to a flow, even if the user hasn't filled out a `deployment.yaml` file.

With this PR, the `prefect deploy` command will prompt the user for a deployment name if one is not provided via the CLI or the projects `deployment.yaml` file. 

The `prefect deploy` command will also prompt the user if a work pool name is not provided. The prompt will show the work pools that are available for the current workspace. Any work pools that are a `prefect-agent` type will not be shown because `prefect deploy` can only be used with typed work pools. If the user doesn't have any work pools that can be used for the deployment, the CLI will guide the user through creating a new work pool.

#### Example output of deploying a flow with only a flow entrypoint given and no configuration in `deployment.yaml`.
Screenshot 2023-05-24 at 9 35 53 AM

#### Videos showing different flows through the experience
[Interactive Deployment w/ Existing Work Pools - Watch Video](https://www.loom.com/share/a79b196707784dd5a3df1c5c1b477c69)

[Interactive Deployment w/o Existing Typed Work Pools - Watch Video](https://www.loom.com/share/3672b84373364c2899585540f6a90bd5)

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.

### CODE REVIEW

**Comment by madkinsz** [comment]
@desertaxle panic on my first attempt

```
❯ prefect deploy flows/hello_world.py:hello
No deployment.yaml file found, only provided CLI options will be used.
Traceback (most recent call last):
  File "/Users/mz/dev/prefect/src/prefect/cli/_utilities.py", line 45, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/mz/dev/prefect/src/prefect/utilities/asyncutils.py", line 260, in coroutine_wrapper
    return call()
           ^^^^^^
  File "/Users/mz/dev/prefect/src/prefect/_internal/concurrency/calls.py", line 245, in __call__
    return self.result()
           ^^^^^^^^^^^^^
  File "/Users/mz/dev/prefect/src/prefect/_internal/concurrency/calls.py", line 173, in result
    return self.future.result(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mz/.pyenv/versions/3.11.2/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/mz/.pyenv/versions/3.11.2/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/mz/dev/prefect/src/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
             ^^^^^^^^^^
  File "/Users/mz/dev/prefect/src/prefect/cli/deploy.py", line 194, in deploy
    with open("prefect.yaml", "r") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'prefect.yaml'
An exception occurred.
```

**Comment by desertaxle** [comment]
> @desertaxle panic on my first attempt

Ah yeah, `prefect deploy` still has an implicit dependency on `prefect project init`. I'll clean up that error.

**Comment by madkinsz** [comment]
In our repository, I did
```
❯ prefect deploy flows/hello_world.py:hello
What would you like to name this deployment?: foo
```

and it hung for a while with no output then began streaming a Docker build

```
What would you like to name this deployment?: foo
Step 1/48 : ARG PYTHON_VERSION=3.8
Step 2/48 : ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
Step 3/48 : ARG BUILD_PYTHON_VERSION=3.8
Step 4/48 : ARG NODE_VERSION=16.15
Step 5/48 : ARG EXTRA_PIP_PACKAGES=""
Step 6/48 : FROM node:${NODE_VERSION}-bullseye-slim as ui-builder
```

I expect there to be more output guiding me through what's happening

**Comment by madkinsz** [comment]
I'd try to base all interactive prompts off the `gh` CLI — for example see `gh pr create`. We did this for `prefect cloud login` as another example. I had to read "What would you like to name this deployment?" a couple times to understand what it was asking me for and the shorter form "? Deployment name" might be clearer.

**Comment by desertaxle** [comment]
> I expect a prompt asking me which recipe I want to use, I think. Or at least confirmation. My `Dockerfile` is not actually needed for running this flow and I want to use `git` but I'm not sure how!

Ooh yeah, a prompt for recipe in `prefect project init` is a good idea!

File: src/prefect/cli/deploy.py [hunk 1]
```diff
@@ -294,7 +334,9 @@ async def _run_single_deploy(
     if not flow_name and not entrypoint:
         raise ValueError("An entrypoint or flow name must be provided.")
     if not name:
-        raise ValueError("A deployment name must be provided.")
+        if not is_interactive():
```
Comment (jakekaplan) [inline src/prefect/cli/deploy.py:337]: Is it worth moving this down until its needed?

```
(py311) ➜  deploy_test git:(interactive-deploy) ✗ prefect deploy helloasdfw.py:main
What would you like to name this deployment?: my_deployment

~~error from register flow cause `helloasdfw` is not a real file~~
```

Comment (jakekaplan) [reply src/prefect/cli/deploy.py:337]: woops meant to leave this lower for `name = prompt("Deployment name")`

**Comment by desertaxle** [comment]
@madkinsz @jakekaplan I believe I've addressed all your comments (except for an interactive prompt for `prefect project init`, which I'll handle in a separate PR). Let me know if you find any other areas that need smoothing out!

**Review by jakekaplan — APPROVED** [review]
🎉

**Pull request merged**